rfcs/rfc-metaprogramming.md

```diff
@@ -0,0 +1,379 @@
+
+Originally, Perl provided some abilities in the form of operations on glob references. Gradually over time, authors have required more abilities than this and increasingly have used a variety of CPAN modules to extend these abilities. As many of the operations provided ought to be considered fundamental language abilities for a dynamic language, it would be better if these were provided by core Perl itself.
+
+This RFC aims to provide a set of API functions to cover many of these abilities. Similar to the case of RFC 0009 having added a namespace for general "builtin" functions, it is not intended that this RFC covers a total and complete list; but instead sets out an initial set of functions. It is expected that more would be added over time on an individual basis, as the need arises. In particular, this RFC does not address any additional API that may be needed by the new object system currently being developed under the `use feature 'class'` flag.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+As this RFC adds a number of functions all on an overall-similar theme, they are all collected into a new toplevel namespace. For now this namespace is called `meta::`; though as per the rest of the details in this RFC that is currently very much up for debate.
+
+Objects that are returned by these functions will be instances of some named package within the `meta::` namespace, though we do not yet specify exactly what they will be. Furthermore we make no statement about what actual container type of objects those actually are. Users should not attempt to interact with these objects except via the API specified here.
+
+### Functions to operate on Packages
+
+These are functions to create, manipulate, and inspect packages themselves.
+
+#### `has_package`
+
+```perl
+$bool = meta::has_package($name);
+```
+
+Returns true if the named package exists.
+
+For example:
+
+```perl
+my $bool = meta::has_package("Example::Package");
+# behaves as
+my $bool = do { no strict 'refs'; exists ${"Example::"}{"Package::"} };
+```
+
+#### `get_package`
+
+```perl
+$metapackage = meta::get_package($name);
+```
+
+Returns a meta-package object instance to represent the named package, if such a package exists. If not an exception is thrown.
+
+#### `add_package`
+
+```perl
+$metapackage = meta::add_package($name);
+```
+
+Creates a new package of the given name, and returns a meta-package object instance to represent it.
+
+#### `list_packages`
+
+```perl
+@metapackages = meta::list_packages($basename);
+```
+
+Returns a list of sub-packages within the given base package name, each as a meta-package object instance.
```

Comment:
If there are none, would this return an empty list, or throw an exception?   The action should be consistent with behaviour of `get_symbol` (whether `undef` or an excpetion) as raised by Ovid there.

Reply:
I'm inclined to differ. When you ask to "get a thing" you really expect to receive that thing, or die trying. Whereas, when you ask to list the possible things, it's quite valid to expect you might end up with nothing. List-returning methods can yield an empty list, whereas conceptually a scalar-returning method must return exactly one thing.

Reply:
A lack of sub-packages isn't exceptional, consider if you were recursively walking a package tree.

Requesting a specific symbol that doesn't exist might be though.

Reply:
This seems a useful pattern though, and even with `use feature "try"`, it's neater when `get_variable` returns undef instead of throwing an exception.

```perl
my $variable = get_variable('$foo') // add_variable('$foo');
```